### ISSUE TITLE
[bug]: lnd (master 15d2ff0) is stalling connections at startup using `externalhosts`

### ISSUE THREAD

**Issue by blckbx** [lightningnetwork/lnd#7924]
### Background

Follow-up issue of #7921 

Good news first: `externalip`  is working now (with and without Tor).
Bad news:  using `externalhosts` LND still stalls connections and lnd becomes unresponsive with ~~and without~~ Tor.

- Compiled master branch from source (15d2ff0).
- LND now starts up without go runtime error but stalls after a while trying to reconnect to peers.
- seeing all connections (in and out) timing out
```
[INF] PEER: Peer(x): disconnecting x@x.x.x.x:37298, reason: unable to start peer: unable to send init msg: write tcp 10.151.111.76:9738->x.x.x.x:58556: write: broken pipe

[ERR] CMGR: Can't accept connection: unable to accept connection from x.x.x.x:56404: read tcp 10.151.111.76:9738->x.x.x.x:56404: i/o timeout
```
- local `lncli` commands are getting stuck and rpc calls are running into timeouts, finally getting cancelled.

Startup log show that lnd is resolving the DNS correctly:
```
[INF] NANN: HostAnnouncer starting
[DBG] NANN: IP change detected! lightning.airdns.org:9738: <nil> -> 195.206.105.203:9738

[DBG] NANN: HostAnnouncer checking for any IP changes...
[DBG] NANN: No IP changes detected for hosts: [lightning.airdns.org:9738]
```

### Your environment

* version of `lnd`: v0.17.0-beta.rc1-g15d2ff0c4
* which operating system (`uname -a` on *Nix): Linux node 5.15.0-79-generic 86-Ubuntu SMP Mon Jul 10 16:07:21 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
* version of `btcd`, `bitcoind`, or other backend: bitcoind v25.0
* any other relevant environment details: using hybrid mode (Tor + externalhosts) setting in combination with vpn service

### Steps to reproduce

- compile 15d2ff0 from source
- `lnd.conf`: set a DNS address for externalhosts
- startup LND
- LND comes to halt trying to connect to peers and stops responding to rpc calls

### Expected behaviour

LND is starting up without issues and responding to calls.

### Actual behaviour

LND starts up with  `externalhosts` set but comes to halt and stops responding to `lncli` commands.

I can provide a full log for externalhosts and externalip scenario privately, if required.

**Comment by Roasbeef** [comment]
Are you able to repro this _without_ Tor? Just tried to repro and seems fine. 

Re Tor, seems related to this issue perhaps https://github.com/lightningnetwork/lnd/issues/7917#issuecomment-1692938283? Might be an intermittent Tor issue. I can try it out on my hybrid tor node once we get the next rc out (rc2).

**Comment by Roasbeef** [comment]
> using hybrid mode (Tor + externalhosts) 

What tor flags are you running with? `lnd.tor.skip-proxy-for-clearnet-targets=1` as well?

**Comment by blckbx** [comment]
> Are you able to repro this _without_ Tor? Just tried to repro and seems fine.

Yes, I ran into this with and without Tor.
With active Tor, `tor.skip-proxy-for-clearnet-targets=true` was set, yes.

**Comment by Roasbeef** [comment]
> So the only problem left seems to be: running clearnet-only with externalhosts setting.

Next time things stall, do you think you can grab a profile? https://github.com/lightningnetwork/lnd/blob/master/docs/debugging_lnd.md#capturing-pprof-data-with-lnd

I ran with tor hybrid and external host and wasn't able to repro.

**Comment by blckbx** [comment]
> I ran with tor hybrid and external host and wasn't able to repro.

This case is not a problem. If  Tor is turned off completely (no hybrid mode, clearnet only), then LND is stalling connections.

[pprof.txt](https://github.com/lightningnetwork/lnd/files/12462101/pprof.txt)
![pprof](https://github.com/lightningnetwork/lnd/assets/74455114/69fbd12d-7467-43a6-946a-ab93a40cfd74)

**Comment by Roasbeef** [comment]
@blckbx can you get a goroutine dump from that? https://go.dev/blog/pprof

Then we can see what is actually blocked.

**Comment by blckbx** [comment]
> There is a goroutine dump here: https://github.com/lightningnetwork/lnd/files/12462101/pprof.txt, though it's in a bit of a different format from what a goroutine dump normally looks like.

Sorry I'm new to this tool. This I copied from the curl command mentioned in debugging guide: `$  curl http://localhost:9736/debug/pprof/goroutine?debug=1`. 

What exact command do I need to run to get the output you can work with?

**Comment by guggero** [comment]
No this is fine, thanks.

**Comment by Roasbeef** [comment]
We have a candidate fix here: https://github.com/lightningnetwork/lnd/issues/7928

**Comment by mercurytoxic** [comment]
Pull request https://github.com/lightningnetwork/lnd/pull/7938 fixed all stalling issues for me.

**Comment by blckbx** [comment]
> We have a candidate fix here: #7928

This indeed fixes the issue with node running clearnet-only using externalhosts setting and thus solves #7924.
